raiden/network/transport/matrix/transport.py

```diff
@@ -133,36 +140,43 @@ def _expiration_generator(
             while now() < _next:  # yield False while next is still in the future
                 yield False
 
-    def enqueue(self, queue_identifier: QueueIdentifier, message: Message) -> None:
+    def enqueue(self, queue_identifier: QueueIdentifier, messages: List[Message]) -> None:
         """ Enqueue a message to be sent, and notify main loop """
         assert queue_identifier.recipient == self.receiver
+
         with self._lock:
-            already_queued = any(
-                queue_identifier == data.queue_identifier and message == data.message
-                for data in self._message_queue
-            )
-            if already_queued:
-                self.log.warning(
-                    "Message already in queue - ignoring",
-                    receiver=to_checksum_address(self.receiver),
-                    queue=queue_identifier,
-                    message=redact_secret(DictSerializer.serialize(message)),
-                )
-                return
             timeout_generator = timeout_exponential_backoff(
                 self.transport._config.retries_before_backoff,
                 self.transport._config.retry_interval_initial,
                 self.transport._config.retry_interval_max,
             )
-            expiration_generator = self._expiration_generator(timeout_generator)
-            self._message_queue.append(
-                _RetryQueue._MessageData(
-                    queue_identifier=queue_identifier,
-                    message=message,
-                    text=MessageSerializer.serialize(message),
-                    expiration_generator=expiration_generator,
+
+            encoded_messages = list()
+            for message in messages:
+                already_queued = any(
+                    queue_identifier == data.queue_identifier and message == data.message
+                    for data in self._message_queue
                 )
-            )
+
+                if already_queued:
+                    self.log.warning(
+                        "Message already in queue - ignoring",
+                        receiver=to_checksum_address(self.receiver),
+                        queue=queue_identifier,
+                        message=redact_secret(DictSerializer.serialize(message)),
+                    )
+                else:
+                    expiration_generator = self._expiration_generator(timeout_generator)
```

Comment:
Here it is very important to **not** share the expiration generator, otherwise we introduced unnecessary delay (currently the generator is used once per message).